doc/proposals/fast_ipam_by_node.md

```diff
@@ -0,0 +1,268 @@
+# Whereabouts support for fast IPAM by using preallocated node slices
+
+# Table of contents
+
+- [Introduction](#introduction)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Design](#design)
+  - [Changes in IPAM Config](#changes-in-ipam-config)
+  - [Changes in Modules](#changes-in-modules)
+  - [Backward compatibility](#backward-compatibility)
+- [Alternative Design](#alternative-design)
+- [Summary](#summary)
+- [Discussions and Decisions](#discussions-and-decisions)
+
+<hr>
+
+## Introduction
+
+Whereabouts currently uses a single lease per cluster named "whereabouts" for locking for all allocation and deallocation
+of IPs across the entire cluster running whereabouts. This causes issues with performance and reliability at scale. 
+Even at 256 nodes, if you have 10 network-attachment-definitions per pod and run a pod on each node there will be so much lease contention that kubelet times
+out before whereabouts can assign all 10 IPs per pod. This would only get worse at higher scale and whereabouts should be able to support
+10+ network-attachment-definition per pod at the kubernetes supported scale of 5,000 nodes.
+
+### Goals
+
+- Support existing whereabouts functionality without breaking changes
+- Introduce a new mode that can be configured on NAD to use IPAM by node slices
+- Support multiple NADs on same network range
+
+### Non-Goals
+
+- Migrate all users to this new mode
+
```

Comment:
Potentially we can add here "integrate reconciliation controller and the new controller" as a non-goal -- that can be left as a follow on if it adds too much overhead. I'm ok with starting with a brand new controller for this process.

Reply:
yes, I was actually thinking about adding this as a non-goal but we may want to do it in the future (with changes to the existing controller) so I left it as is for now but I can add it